From 6cf4d8bf635a69d1bd714758011b541f53f931ec Mon Sep 17 00:00:00 2001
From: Yong Zhang <yong.zhang@windriver.com>
Date: Mon, 30 Aug 2010 02:10:57 -0700
Subject: [PATCH] MIPS: OCTEON: restore saved K0 if OCTEON_FAST_PATH_TLB_REFILL_HANDLER_SAVEK0 is enabled

When OCTEON_FAST_PATH_TLB_REFILL_HANDLER_SAVEK0 is enabled, K0 is saved
to prevent it from being mangled. But it's never restored. This
will lead to some error when we do vmalloc stress test.

Signed-off-by: Yong Zhang <yong.zhang@windriver.com>
---
 arch/mips/mm/tlbex.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/arch/mips/mm/tlbex.c b/arch/mips/mm/tlbex.c
index 7339795..1e756d8 100644
--- a/arch/mips/mm/tlbex.c
+++ b/arch/mips/mm/tlbex.c
@@ -705,6 +705,9 @@ build_get_pgd_vmalloc64(u32 **p, struct uasm_label **l, struct uasm_reloc **r,
            // branch to large_segbits_fault if xuseg or xsseg address
            uasm_il_bgez(p, r, bvaddr, label_large_segbits_fault);
 #endif
+#ifdef OCTEON_FAST_PATH_TLB_REFILL_HANDLER_SAVEK0
+	uasm_i_move(p, USER_REGISTER_TO_USE, bvaddr);
+#endif
         }
 
 #ifdef MODULE_START
-- 
1.7.0.4

